flow/environment.py

```diff
@@ -90,6 +91,169 @@ def template_filter(func):
     return classmethod(func)
 
 
+class _NodeTypes(enum.Enum):
+    """Defines for a partition the acceptable node requests.
+
+    - SHARED: Only support partial nodes requests to full node. Note that you
+        can restrict the cores provided in `_PartitionConfig` to restrict below
+        even full nodes.
+    - MIXED: No restriction on partial to full node requests.
+    - WHOLENODE: Only support submissions in units of whole nodes.
+    """
+
+    SHARED = 1
+    MIXED = 2
+    WHOLENODE = 3
+
+
+class _PartitionConfig:
+    """The configuration of partition for a given environment.
+
+    Currently supports the ideas of
+    - CPUs for a partition
+    - GPUs for a partition
+    - Node type of a partition
+
+    When querying a value for a specific partition, the logic first searches the
+    provided mapping, if any, for the partition. If it is not found, then the
+    mapping is searched for "default" if it exists. If not, the class default is
+    used.
+
+    1. Partition specific -> 2. Provided default -> 3. _PartitionConfig default
+
+    The class defaults are
+
+    - CPUs: ``None`` which represents any number. This will be interpretted as
+      an unlimited supply of CPUs practically.
+    - GPUs: 0.
+    - Node type: `_NodeTypes.MIXED`.
+
+    Parameters
+    ----------
+    cpus_per_node: dict[str, int], optional
+        Mapping between partitions and CPUs per node. Defaults to an empty `dict`.
```

Comment:
I had to look at the environments to understand this docstring. Need to mention that keys in this dictionary are partition names.

I understand why "mapping" was used, but I couldn't figure out which direction it was pointing and how to populate the dictionary based on this info.

```suggestion
        A `dict` mapping from partition names (keys) to CPUs per node (values) on those partitions.
        Defaults to an empty `dict`.
```

Also applies to the other docstrings but we can discuss here first

Reply:
A mapping from a to b means a are keys and b are values.